modules/redcap/php/redcapmapper.class.inc

```diff
@@ -0,0 +1,467 @@
+
+        $site_name = $visit_mapping->create_session->site_name;
+        $site_id   = array_search($site_name, \Utility::getSiteList());
+
+        $project_name = $visit_mapping->create_session->project_name;
+        $project_id   = array_search($project_name, \Utility::getProjectList());
+
+        $cohort_name = $visit_mapping->create_session->cohort_name;
+        $cohort_id   = array_search($cohort_name, \Utility::getCohortList());
+
+        // Throw an exception if any of the site, project, or cohort could not be
+        // obtained from the database, which means that the information present in
+        // the session creation configuration was incorrect.
+
+        if (!$site_id) {
+            throw new \LorisException(
+                "[redcap] Error: No LORIS site found for site name '$site_name',"
+                . " the REDCap module configuration is incorrect."
+            );
+        }
+
+        if (!$project_id) {
+            throw new \LorisException(
+                "[redcap] Error: No LORIS project found for site name"
+                . " '$project_name', the REDCap module configuration is incorrect."
+            );
+        }
+
+        if (!$cohort_id) {
+            throw new \LorisException(
+                "[redcap] Error: No LORIS cohort found for site name"
+                . " '$cohort_name', the REDCap module configuration is incorrect."
+            );
+        }
+
+        // Since the REDCap module is called by the REDCap API or from a script,
+        // there is usually no logged in user. As such, manually set the user to be
+        // the REDCap issue assignee to create the session.
+
+        $user = $this->_queries->getRedcapIssueAssignee();
+
+        $state = \State::singleton();
+        $state->setUsername($user->getUsername());
+        $_SESSION['State'] =& $state;
+
+        // Create the session that corresponds to this mapping.
+        \TimePoint::createNew(
+            $candidate,
+            $cohort_id,
+            $visit_mapping->visit_label,
+            \Site::singleton(new \CenterID(strval($site_id))),
+            \Project::getProjectFromID(new \ProjectID(strval($project_id))),
+        );
+
+        // Get the session again now that it been created.
+
+        $session_id = array_search(
+            $visit_mapping->visit_label,
+            $candidate->getListOfVisitLabels()
+        );
+
+        $session = \TimePoint::singleton(new \SessionID(strval($session_id)));
+
+        // Start the session next stage, so that it can receive instrument data.
+
+        $new_stage = $session->getNextStage();
+        $session->startStage($new_stage);
+        // TODO: Ideally, we should use the REDCap record date instead of the REDCap
+        // notification date.
+        $date_string = $redcap_notif->received_datetime->format('Y-m-d');
+        $session->setData(
+            [
+                "Date_{$new_stage}" => $date_string,
+            ]
+        );
```

Comment:
2 cents here: the record date should be inside the data, I am not sure if that works the same way with C-BIG REDCap instances, but the REDCap field name should be `timestamp`, which correspond to the end server save timestamp on an instrument (when someone clicks the REDCap form final "save values" button).